### PR TITLE
Add more return types for visitor functions

### PR DESCRIPTION
**Pull request by abhinavmehndiratta**
The visitor return value names are not final and yet to discussed. I will update the documentation accordingly.
I created a table for comparing the benchmarks. There is a slight performance hit (5-10%). The number of memory allocations are unaffected.
```
julia> g = loadsnap(:soc_slashdot0902_u)
{82168, 582533} undirected simple Int64 graph

julia> dg = loadsnap(:ego_twitter_d)
{81306, 1768149} directed simple Int64 graph
```

**BENCHMARKS**
| Test                                                  | New time   | Old time   |
|-------------------------------------------------------|------------|------------|
| connected_components(g, DFS())                        | 28.892 ms  | 26.824 ms  |
| connected_components(dg, Kosaraju())                  | 55.429 ms  | 50.502 ms  |
| connected_components(dg, Tarjan())                    | 36.570 ms  | 34.143 ms  |
| distances(g, ss, DepthFirst())                        | 27.230 ms  | 25.082 ms  |
| distances(g, ss, BreadthFirst())                      | 10.846 ms  | 9.662 ms   |
| distances(g, ss, ThreadedBreadthFirst())              | 3.117 ms   | 2.975 ms   |
| parents(g, ss, DepthFirst())                          | 26.929 ms  | 24.852 ms  |
| parents(g, ss, BreadthFirst())                        | 10.449 ms  | 9.654 ms   |
| parents(g, ss, ThreadedBreadthFirst())                | 3.125 ms   | 2.989 ms   |
| distances(dg, ss, DepthFirst())                       | 25.672 ms  | 23.866 ms  |
| distances(dg, ss, BreadthFirst())                     | 15.784 ms  | 14.215 ms  |
| distances(dg, ss, ThreadedBreadthFirst())             | 4.767 ms   | 4.526 ms   |
| parents(dg, ss, DepthFirst())                         | 25.535 ms  | 23.934 ms  |
| parents(dg, ss, BreadthFirst())                       | 15.784 ms  | 14.458 ms  |
| parents(dg, ss, ThreadedBreadthFirst())               | 4.783 ms   | 4.538 ms   |
| shortest_paths(g, ss, TrackingBFS())                  | 30.665 ms  | 29.654 ms  |
| shortest_paths(dg, ss, TrackingBFS())                 | 33.841 ms  | 31.971 ms  |
| ____has_path on 100 pairs of (u, v)____               |            |            |
| test_has_path(g, DepthFirst(), srcs, dsts)            | 1.315 s    | 1.233 s    |
| test_has_path(g, BreadthFirst(), srcs, dsts)          | 262.609 ms | 246.744 ms |
| test_has_path(g, ThreadedBreadthFirst(), srcs, dsts)  | 192.383 ms | 185.362 ms |
| test_has_path(dg, DepthFirst(), srcs, dsts)           | 811.993 ms | 750.751 ms |
| test_has_path(dg, BreadthFirst(), srcs, dsts)         | 349.377 ms | 318.146 ms |
| test_has_path(dg, ThreadedBreadthFirst(), srcs, dsts) | 198.299 ms | 187.514 ms |

I've also corrected some errors which I encountered on testing LightGraphs locally. Since we don't have multi-threading enabled on CI, it was unable to detect them.

Here is the benchmark code for reference:
```julia
julia> using LightGraphs

julia> using LightGraphs.Connectivity: connected_components, DFS, Kosaraju, Tarjan

julia> using LightGraphs.Traversals: distances, parents, has_path, BreadthFirst, DepthFirst, ThreadedBreadthFirst

julia> using LightGraphs.ShortestPaths: shortest_paths, TrackingBFS

julia> using SNAPDatasets, BenchmarkTools

julia> using Random

julia> Random.seed!(1);

julia> BenchmarkTools.DEFAULT_PARAMETERS.samples = 100
100

julia> Threads.nthreads()
6

julia> g = loadsnap(:soc_slashdot0902_u)
{82168, 582533} undirected simple Int64 graph

julia> dg = loadsnap(:ego_twitter_d)
{81306, 1768149} directed simple Int64 graph

julia> # wcc
       @btime connected_components(g, DFS())
  28.892 ms (30 allocations: 3.89 MiB)

julia> # scc
       @btime connected_components(dg, Kosaraju())
  55.429 ms (24917 allocations: 7.64 MiB)

julia> @btime connected_components(dg, Tarjan())
  36.570 ms (24586 allocations: 8.25 MiB)

julia> # undirected distances and parents
       ss = [1, 5, 6];

julia> @btime distances(g, ss, DepthFirst())
  27.230 ms (20 allocations: 1.64 MiB)

julia> @btime distances(g, ss, BreadthFirst())
  10.846 ms (9 allocations: 1.89 MiB)

julia> @btime distances(g, ss, ThreadedBreadthFirst())
  3.117 ms (443 allocations: 2.00 MiB)

julia> @btime parents(g, ss, DepthFirst())
  26.929 ms (20 allocations: 1.64 MiB)

julia> @btime parents(g, ss, BreadthFirst())
  10.449 ms (9 allocations: 1.89 MiB)

julia> @btime parents(g, ss, ThreadedBreadthFirst())
  3.125 ms (443 allocations: 2.00 MiB)

julia> # directed distances and parents
       ss = [1, 7, 10];

julia> @btime distances(dg, ss, DepthFirst())
  25.672 ms (20 allocations: 1.63 MiB)

julia> @btime distances(dg, ss, BreadthFirst())
  15.784 ms (9 allocations: 1.87 MiB)

julia> @btime distances(dg, ss, ThreadedBreadthFirst())
  4.767 ms (401 allocations: 1.98 MiB)

julia> @btime parents(dg, ss, DepthFirst())
  25.535 ms (20 allocations: 1.63 MiB)

julia> @btime parents(dg, ss, BreadthFirst())
  15.883 ms (9 allocations: 1.87 MiB)

julia> @btime parents(dg, ss, ThreadedBreadthFirst())
  4.783 ms (399 allocations: 1.98 MiB)

julia> # shortest paths tracking bfs
       ss = [1, 50, 200];

julia> @btime shortest_paths(g, ss, TrackingBFS())
  30.665 ms (133869 allocations: 15.43 MiB)

julia> @btime shortest_paths(dg, ss, TrackingBFS())
  33.841 ms (147797 allocations: 16.69 MiB)

julia> # has path on 100 pairs of (u, v)
       function test_has_path(g, alg, srcs, dsts) where T
               for i in 1:100
                       has_path(g, srcs[i], dsts[i], alg)
               end
       end
test_has_path (generic function with 1 method)

julia> srcs = rand(1:nv(g), 100);

julia> dsts = rand(1:nv(g), 100);

julia> @btime test_has_path(g, DepthFirst(), srcs, dsts)
  1.315 s (2257 allocations: 86.50 MiB)

julia> @btime test_has_path(g, BreadthFirst(), srcs, dsts)
  262.609 ms (1200 allocations: 126.43 MiB)

julia> @btime test_has_path(g, ThreadedBreadthFirst(), srcs, dsts)
  192.383 ms (27656 allocations: 142.10 MiB)

julia> srcs = rand(1:nv(dg), 100);

julia> dsts = rand(1:nv(dg), 100);

julia> @btime test_has_path(dg, DepthFirst(), srcs, dsts)
  811.993 ms (2017 allocations: 64.35 MiB)

julia> @btime test_has_path(dg, BreadthFirst(), srcs, dsts)
  349.377 ms (1200 allocations: 125.11 MiB)

julia> @btime test_has_path(dg, ThreadedBreadthFirst(), srcs, dsts)
  198.299 ms (27254 allocations: 134.39 MiB)
```

### CODE REVIEW

**Comment by abhinavmehndiratta** [comment]
< 5% slowdown after using `@inline` and `UInt8`

**UPDATED BENCHMARKS**
| Test                                                  | New time   | Old time   |
|-------------------------------------------------------|------------|------------|
| connected_components(g, DFS())                        | 26.905 ms  | 26.824 ms  |
| connected_components(dg, Kosaraju())                  | 52.362 ms  | 50.502 ms  |
| connected_components(dg, Tarjan())                    | 35.447 ms  | 34.143 ms  |
| distances(g, ss, DepthFirst())                        | 25.109 ms  | 25.082 ms  |
| distances(g, ss, BreadthFirst())                      | 10.200 ms  | 9.662 ms   |
| distances(g, ss, ThreadedBreadthFirst())              | 2.984 ms   | 2.975 ms   |
| parents(g, ss, DepthFirst())                          | 24.950 ms  | 24.852 ms  |
| parents(g, ss, BreadthFirst())                        | 10.116 ms  | 9.654 ms   |
| parents(g, ss, ThreadedBreadthFirst())                | 3.125 ms   | 2.989 ms   |
| distances(dg, ss, DepthFirst())                       | 24.316 ms  | 23.866 ms  |
| distances(dg, ss, BreadthFirst())                     | 15.241 ms  | 14.215 ms  |
| distances(dg, ss, ThreadedBreadthFirst())             | 4.563 ms   | 4.526 ms   |
| parents(dg, ss, DepthFirst())                         | 24.416 ms  | 23.934 ms  |
| parents(dg, ss, BreadthFirst())                       | 15.266 ms  | 14.458 ms  |
| parents(dg, ss, ThreadedBreadthFirst())               | 4.563 ms   | 4.538 ms   |
| shortest_paths(g, ss, TrackingBFS())                  | 29.848 ms  | 29.654 ms  |
| shortest_paths(dg, ss, TrackingBFS())                 | 32.641 ms  | 31.971 ms  |
| ____has_path on 100 pairs of (u, v)____               |            |            |
| test_has_path(g, DepthFirst(), srcs, dsts)            | 1.275 s    | 1.233 s    |
| test_has_path(g, BreadthFirst(), srcs, dsts)          | 260.609 ms | 246.744 ms |
| test_has_path(g, ThreadedBreadthFirst(), srcs, dsts)  | 187.383 ms | 185.362 ms |
| test_has_path(dg, DepthFirst(), srcs, dsts)           | 777.993 ms | 750.751 ms |
| test_has_path(dg, BreadthFirst(), srcs, dsts)         | 334.377 ms | 318.146 ms |
| test_has_path(dg, ThreadedBreadthFirst(), srcs, dsts) | 192.529 ms | 187.514 ms |

**Comment by abhinavmehndiratta** [comment]
@sbromberger it seems codecov isn't working

File: src/Traversals/Traversals.jl [hunk 1]
```diff
@@ -319,9 +334,13 @@ end
 function preinitfn!(s::PathState, visited)
     visited[s.exclude_vertices] .= true
     s.vertices_in_exclude = visited[s.u] | visited[s.v]
-    return !s.vertices_in_exclude
+    !s.vertices_in_exclude && return VSUCCESS
```
Comment (sbromberger) [inline src/Traversals/Traversals.jl:337]: I would like these to simplified where possible. That is,
```
return !s.vertices_in_exclude ? VSUCCESS : VTERMINATE
```

Where we can't simplify, the "happy path" (i.e., the one that leads to VSUCCESS) should be the normal exit.

File: src/Traversals/depthfirst.jl [hunk 2]
```diff
@@ -19,42 +43,77 @@ function traverse_graph!(
...
             v, _ = S[end]
-            previsitfn!(state, v) || return false
-
+            is_successful(previsitfn!(state, v)) || return false
             neighs=alg.neighborfn(g, v)
             @inbounds while ptr <= length(neighs)
                 i = neighs[ptr]
-                visitfn!(state, v, i) || return false
+                x = visitfn!(state, v, i)
+                x ==  VTERMINATE && return false # terminate dfs
```
Comment (sbromberger) [inline src/Traversals/depthfirst.jl:64]: spacing?

File: src/Traversals/depthfirst.jl [hunk 3]
```diff
@@ -19,42 +43,77 @@ function traverse_graph!(
...
+                if x == VFAIL # stop exploring all neighbors
+                    ptr = length(neighs)+1
+                    break
+                end
                 if !visited[i]  # find the first unvisited neighbor
-                    newvisitfn!(state, v, i) || return false
+                    # newvisitfn! return values have same effect as visitfn! but only for
+                    # newly discovered vertices
+                    x = newvisitfn!(state, v, i)
+                    x ==  VTERMINATE && return false
```
Comment (sbromberger) [inline src/Traversals/depthfirst.jl:77]: spacing?

File: src/Traversals/depthfirst.jl [hunk 4]
```diff
@@ -19,42 +43,77 @@ function traverse_graph!(
...
+                    end
                     visited[i] = true
                     push!(S, (i, ptr+1))
                     break
                 else
-                    revisitfn!(state, v, i) || return false
+                    # revisitfn! return values have same effect as visitfn! but only for
+                    # rediscovered vertices
+                    x = revisitfn!(state, v, i)
+                    x ==  VTERMINATE && return false
```
Comment (sbromberger) [inline src/Traversals/depthfirst.jl:93]: spacing?

File: src/Traversals/threaded-breadthfirst.jl [hunk 5]
```diff
@@ -9,6 +9,21 @@ at a time (default `20`). For graphs with uniform degree, a larger value of
 `queue_segment_size` may improve performance.
 - `neighborfn::Function`: the function to use to compute the neighbors of a vertex (default [`outneighbors`](@ref)).
 
+### Visitor Function Usage
+- `preinitfn!`: continue with normal execution if `VSUCCESS`, otherwise terminate
+- `initfn!`: continue with normal execution if `VSUCCESS`, otherwise terminate
+- `previsitfn!`: continue with normal execution if `VSUCCESS`, otherwise terminate
+- `visitfn!`: this function is called for every neighbor (irrespective of whether it is already
+              visited or not) of the vertex being explored
+              if `VTERMINATE`: terminate complete traversal
```
Comment (sbromberger) [inline src/Traversals/threaded-breadthfirst.jl:18]: bullets needed here.

File: src/Traversals/threaded-breadthfirst.jl [hunk 6]
```diff
@@ -105,28 +120,41 @@ function traverse_graph!(
                         visited[v] || continue
                         # (visited[v] && vert_level[v] == n_level-one(U)) || continue
                         for i in alg.neighborfn(g, v)
-                            if !visitfn!(state, v, i, t)
+                            x = visitfn!(state, v, i, thread_id)
+                            if x ==  VTERMINATE # terminate bfs
```
Comment (sbromberger) [inline src/Traversals/threaded-breadthfirst.jl:124]: spacing?

File: src/Traversals/threaded-breadthfirst.jl [hunk 7]
```diff
@@ -105,28 +120,41 @@ function traverse_graph!(
...
                             end
+                            x == VSKIP && continue  # skip to next neighbor
+                            x == VFAIL && break # stop exploring curr vertex's neighbors
                             # Data race, but first read on visited[i] always succeeds
                             if !visited[i]
-                                if !newvisitfn!(state, v, i, t)
+                                # newvisitfn! return values have same effect as visitfn! but only for
+                                # newly discovered vertices
+                                x = newvisitfn!(state, v, i, thread_id)
+                                if x ==  VTERMINATE
```
Comment (sbromberger) [inline src/Traversals/threaded-breadthfirst.jl:135]: spacing?

File: src/Traversals/threaded-breadthfirst.jl [hunk 8]
```diff
@@ -105,28 +120,41 @@ function traverse_graph!(
...
                                 # vert_level[i] = n_level
                                 #Concurrent visited[i] = true always succeeds
                                 visited[i] = true
                                 push!(next_level, i)
                             else
-                                if !revisitfn!(state, v, i, t)
+                                # newvisitfn! return values have same effect as visitfn! but only for
+                                # rediscovered vertices
+                                x = revisitfn!(state, v, i, thread_id)
+                                if x ==  VTERMINATE
```
Comment (sbromberger) [inline src/Traversals/threaded-breadthfirst.jl:149]: spacing

File: test/ShortestPaths/threaded-bfs.jl [hunk 9]
```diff
@@ -4,14 +4,16 @@
     add_edge!(g5, 1, 2); add_edge!(g5, 2, 3); add_edge!(g5, 1, 3); add_edge!(g5, 3, 4)
 
     @testset "$g" for g in testdigraphs(g5)
-        @test @inferred(ShortestPaths.shortest_paths(g, 1, ShortestPaths.ThreadedBFS()))  == ShortestPaths.shortest_paths(g, 1, ShortestPaths.BFS())
-        @test @inferred(ShortestPaths.shortest_paths(g, [1, 3], ShortestPaths.ThreadedBFS()))  == ShortestPaths.shortest_paths(g, [1, 3], ShortestPaths.BFS())
+        # ThreadedBFS parents may differ from BFS
+        @test @inferred(ShortestPaths.shortest_paths(g, 1, ShortestPaths.ThreadedBFS())).dists  == ShortestPaths.shortest_paths(g, 1, ShortestPaths.BFS()).dists
+        @test @inferred(ShortestPaths.shortest_paths(g, [1, 3], ShortestPaths.ThreadedBFS())).dists  == ShortestPaths.shortest_paths(g, [1, 3], ShortestPaths.BFS()).dists
```
Comment (sbromberger) [inline test/ShortestPaths/threaded-bfs.jl:9]: this isn't your issue but there's some spacing here.

File: test/ShortestPaths/threaded-bfs.jl [hunk 10]
```diff
@@ -4,14 +4,16 @@
...
     end
 
     g6 = SimpleGraph(SGGEN.House())
 
     @testset "$g" for g in testgraphs(g6)
-        @test @inferred(ShortestPaths.shortest_paths(g, 2, ShortestPaths.ThreadedBFS()))  == ShortestPaths.shortest_paths(g, 2, ShortestPaths.BFS())
-        @test @inferred(ShortestPaths.shortest_paths(g, [1, 2], ShortestPaths.ThreadedBFS())) == ShortestPaths.shortest_paths(g, [1, 2], ShortestPaths.BFS())
+        # ThreadedBFS parents may differ from BFS
+        @test @inferred(ShortestPaths.shortest_paths(g, 2, ShortestPaths.ThreadedBFS())).dists  == ShortestPaths.shortest_paths(g, 2, ShortestPaths.BFS()).dists
+        @test @inferred(ShortestPaths.shortest_paths(g, [1, 2], ShortestPaths.ThreadedBFS())).dists == ShortestPaths.shortest_paths(g, [1, 2], ShortestPaths.BFS()).dists
```
Comment (sbromberger) [inline test/ShortestPaths/threaded-bfs.jl:17]: see above

**Pull request merged**